Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
     return m_currentOpacity < 1;
 }
 
+bool TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective() const
+{
+    // Check if 3D scene leaf or if has perspective and has children and has 3D transformation
+    if (((!m_state.preserves3D && (m_parent && m_parent->preserves3D())) || m_layerTransforms.localTransform.hasPerspective())
+        && !m_children.isEmpty() && has3DLocalTransform())
+            return true;
+    return false;
+}
+
+bool TextureMapperLayer::has3DLocalTransform() const
+{
+    return !(!m_layerTransforms.localTransform.m13() && !m_layerTransforms.localTransform.m23()
+        && !m_layerTransforms.localTransform.m31() && !m_layerTransforms.localTransform.m32()
+        && m_layerTransforms.localTransform.m33() == 1 && !m_layerTransforms.localTransform.m34()
+        && !m_layerTransforms.localTransform.m43() && m_layerTransforms.localTransform.m44() == 1);
+}
+
+bool TextureMapperLayer::accumulatesTransform() const
+{
+    return !m_state.masksToBounds && (preserves3D() || (m_parent && m_parent->preserves3D()));
+}
+
+bool TextureMapperLayer::preserves3D() const
```

Comment:
Candidate for inlining in the header.

Reply:
Inlining causes style check failure "Inline functions should not be in classes annotated with WEBCORE_EXPORT"

Reply:
It sholuld be inlined. The best approach in this case is removing the class-level WEBCORE_EXPORT before this change.

Reply:
I will do that in https://bugs.webkit.org/show_bug.cgi?id=281899

Reply:
ok, now I understand. Sorry for misunderstanding